### PR TITLE
dev-petros - Branch Optimization

### PR DESCRIPTION
**Pull request by ptoupas**

### CODE REVIEW

File: fpgaconvnet/models/network/visualise.py
Comment (Yu-Zhewen) [inline fpgaconvnet/models/network/visualise.py:82]: there is already a "visualise" function in Network.py, and I am not sure how they are different

File: fpgaconvnet/models/network/Network.py
Comment (Yu-Zhewen) [inline fpgaconvnet/models/network/Network.py:204]: nice! I think the name of the function might be a bit too broad or vague, as validity might refer to many things. How about something like "check_graph_completeness"

File: fpgaconvnet/models/partition/update.py
Comment (Yu-Zhewen) [inline fpgaconvnet/models/partition/update.py:48]: what if ports_in is reduced

File: fpgaconvnet/models/partition/update.py
Comment (Yu-Zhewen) [inline fpgaconvnet/models/partition/update.py:43]: where is this function called from? I only found it at "https://github.com/AlexMontgomerie/fpgaconvnet-optimiser/blob/c49f0f5ed6063158c68d799e7652efe162746a3f/fpgaconvnet/optimiser/solvers/solver.py#L350" and it is commented

**Review by Yu-Zhewen — APPROVED** [review]

File: fpgaconvnet/models/network/visualise.py
Comment (ptoupas) [reply fpgaconvnet/models/network/visualise.py:82]: Actually, I created this one before seeing that there was another one doing the same thing.
In any case, the original Network visualise function depends on the visualise functions for each layer of the graph, which are not always implemented and are currently a bit buggy.
I kept this one as I need to check the partitions and how they are structured (mostly as a debugging tool).
We might drop this before merging, BUT we need to fix the other one in that case.

File: fpgaconvnet/models/network/Network.py
Comment (ptoupas) [reply fpgaconvnet/models/network/Network.py:204]: Yeah, that is fair enough. I will rename this one to "check_graph_completeness" as you suggested.

File: fpgaconvnet/models/partition/update.py
Comment (ptoupas) [reply fpgaconvnet/models/partition/update.py:43]: That was an initial attempt to deal with the multiport layers.
This one is not currently being called anywhere.
I kept it there just in case we needed it in the future, but I don't think this is happening.
I am going to completely remove this one

File: fpgaconvnet/models/partition/update.py
Comment (ptoupas) [reply fpgaconvnet/models/partition/update.py:48]: This one is going to be removed after all.
I know it had issues like the one you pointed out but I wasn't actually using that

**Comment by AlexMontgomerie** [comment]
Are we happy to merge this, even if it is failing tests at the moment?

**Pull request merged**
